### PR TITLE
fix: added info about stateless lambdas optimization (KT-32067)

### PR DESCRIPTION
**Pull request by p7nov**
Added a note about creating stateless lambdas as singletons per requests from YT and Discourse:
https://youtrack.jetbrains.com/issue/KT-32067
https://discuss.kotlinlang.org/t/lambdas-and-implicit-references-to-the-instance-of-the-enclosing-class/

### CODE REVIEW

**Comment by ilya-g** [comment]
I'd say that only one instance of that lambda is created without going into details that it's compiled to a singleton class.

File: pages/docs/reference/lambdas.md [hunk 1]
```diff
@@ -430,6 +430,21 @@ print(sum)
...
 
+For lambdas that don't capture any variables (also know as _stateless_ lambdas), there is a built-in optimization:
+the compiler creates only one instance for each such lambda. So, all instances of a stateless lambda point to the same instance.
+
+<div class="sample" markdown="1" theme="idea" data-highlight-only>
+ 
+ ```kotlin
+ val lambdaProvider = {
+    { Unit }  // returns a stateless lambda
+}
```
Comment (ilya-g) [inline pages/docs/reference/lambdas.md]: The example with a lambda stored in a property and returning another Unit-returning lambda is rather complicated to illustrate the concept of stateless lambda.

An alternative could be a function returning either a stateless or stateful lambda:
```kotlin
fun searchPredicate(expected: String? = null): (String?) -> Boolean {
    if (expected == null)
        return { value -> value == null } // stateless
    else
        return { value -> value == expected } // capturing
}
```

File: pages/docs/reference/lambdas.md [hunk 2]
```diff
@@ -430,6 +430,21 @@ print(sum)
 
 </div>
 
+For lambdas that don't capture any variables (also know as _stateless_ lambdas), there is a built-in optimization:
+the compiler creates only one instance for each such lambda. So, all instances of a stateless lambda point to the same instance.
```
Comment (ilya-g) [inline pages/docs/reference/lambdas.md]: > all instances of a stateless lambda point to the same instance.

This confuses the matter: on one hand there is the single instance, on the other—there are many ('all') of them.

**Comment by p7nov** [comment]
The described optimization appears to be a platform-specific optimization and is not a part of the design. So, we shouldn't mention it in the common public doc.

**Comment by udalov** [comment]
Why not just add "On JVM, ..."?

**Comment by p7nov** [comment]
As discussed with authors of the Kotlin language specification (https://github.com/Kotlin/kotlin-spec), this is an implementation detail that can be changed and is not guaranteed in the future. 
Currently, we've added the description of this optimization to the spec as a platform-specific note: https://github.com/Kotlin/kotlin-spec/blob/master/docs/kotlin.jvm/lambda-expressions.md (not sure where they will put it, now it's just a new file with this note).

**Pull request closed without merging**